src/tools/ceph_dedup_tool.cc

```diff
@@ -562,34 +569,68 @@ class SampleDedupWorkerThread : public Thread
       std::unique_lock lock(fingerprint_lock);
       auto found_iter = fp_map.find(chunk.fingerprint);
       ssize_t cur_reference = 1;
+      if (is_hit_threshold()) {
+	auto len_before = fp_map.size();
+	try_shrink();
+	auto len_after = fp_map.size();
+	if (len_before - len_after < len_before / 4) {
+	  state = fp_store_state_t::FULL;
+	  cout << "FpStore reaches memory threshold" << std::endl;
+	}
+      }
       if (found_iter == fp_map.end()) {
-        fp_map.insert({chunk.fingerprint, 1});
+	if (state != fp_store_state_t::FULL) {
```

Comment:
This state member isn't used anywhere else, remove it and use a local instead.  The is_hit_threshhold() check above is entirely unnecessary except in this branch, move here.  Finally, it looks to me like you are doing a full scan of the fp_store every time you try to add a fingerprint once it becomes full -- isn't that incredibly expensive?

Reply:
Moved. In addition, to address a full scan overhead, I added the splitted map data structures---under_threshold_fp_map and over_threshold_fp_map. Using these data structures, Fpstore only clears the under_threshold_fp_map without a full scan under memory pressure. Note that memory usage is the sum of both maps' size.